persistent-postgresql/test/MigrationSpec.hs

```diff
@@ -0,0 +1,404 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module MigrationSpec where
+
+import PgInit
+
+import qualified Data.Map as Map
+import Data.Proxy
+import qualified Data.Text as T
+import Database.Persist.Postgresql.Internal.Migration
+import qualified Database.Persist.SqlBackend.Internal as SqlBackend
+
+getConnPrepare
+    :: (Monad m) => SqlPersistT m (Text -> IO Statement)
+getConnPrepare = do
+    backend <- ask
+    pure (SqlBackend.connPrepare backend)
+
+-- NB: we do not perform these migrations in main.hs
```

Comment:
the test suite is kind of cursed in that it leaves the schema in the test DB, because this means that if there's a bug in the code that determines what migrations you need to apply, you only see test failures if you run the tests twice in a row (if you're starting from a fresh DB).

I think I've made things slightly worse here by doing this... but I also think it's important to directly exercise the migrator like this.